src/Grpc.AspNetCore.Web/Internal/GrpcWebProtocolHelpers.cs

```diff
@@ -0,0 +1,116 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.IO.Pipelines;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    internal static class GrpcWebProtocolHelpers
+    {
+        private const byte Cr = (byte)'\r';
+        private const byte Lf = (byte)'\n';
+        private const byte Colon = (byte)':';
+        private const byte Space = (byte)' ';
+
+        // Special trailers byte with eigth most significant bit set.
+        // Parsers will use this to identify regular messages vs trailers
+        private static readonly byte Trailers = 0x80;
+
+        private static readonly int HeaderSize = 5;
+
+        public static async Task WriteTrailersAsync(IHeaderDictionary trailers, PipeWriter output)
+        {
+            // Flush so the last message is written as its own base64 segment
+            await output.FlushAsync();
+
+            WriteTrailers(trailers, output);
+
+            await output.FlushAsync();
+        }
+
+        internal static void WriteTrailers(IHeaderDictionary trailers, IBufferWriter<byte> output)
+        {
+            // Precalculate trailer size. Required for trailers header metadata
+            var contentSize = CalculateHeaderSize(trailers);
+
+            var totalSize = contentSize + HeaderSize;
+            var buffer = output.GetSpan(totalSize);
+
+            WriteTrailersHeader(buffer, contentSize);
+            WriteTrailersContent(buffer.Slice(HeaderSize), trailers);
+
+            output.Advance(totalSize);
+        }
+
+        private static void WriteTrailersHeader(Span<byte> buffer, int length)
+        {
+            buffer[0] = Trailers;
+            BinaryPrimitives.WriteUInt32BigEndian(buffer.Slice(1), (uint)length);
+        }
+
+        private static int CalculateHeaderSize(IHeaderDictionary trailers)
+        {
+            var total = 0;
+            foreach (var kv in trailers)
+            {
+                foreach (var value in kv.Value)
+                {
+                    if (value != null)
+                    {
+                        // Key + value + 2 (': ') + 2 (\r\n)
+                        total += kv.Key.Length + value.Length + 4;
+                    }
+                }
+            }
+
+            return total;
+        }
+
+        private static void WriteTrailersContent(Span<byte> buffer, IHeaderDictionary trailers)
+        {
+            var currentBuffer = buffer;
+
+            foreach (var kv in trailers)
+            {
+                foreach (var value in kv.Value)
```

Comment:
You haven't actually verified at any point before this that the key and value only contain ASCII, or that there are no prohibited ASCII characters like control characters. The servers each have their own IHeaderDictionary implementation that enforces this.

Reply:
What validation do you suggestion?

Something like this...? https://github.com/aspnet/AspNetCore/blob/4dd9bfc492452174f9329c3e1fc4807cb393c05a/src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs#L257-L276

Reply:
Right.